core/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java

```diff
@@ -1633,26 +1660,61 @@ public void testSeqNoAndCheckpoints() throws IOException {
+                    id = randomFrom(indexedIds);
+                    final Engine.Delete delete = new Engine.Delete(
+                        "test", id, newUid("test#" + id), SequenceNumbersService.UNASSIGNED_SEQ_NO,
+                        rarely() ? 100 : Versions.MATCH_ANY, VersionType.INTERNAL, PRIMARY, 0, false);
+                    try {
+                        initialEngine.delete(delete);
+                        indexedIds.remove(id);
+                    } catch (VersionConflictEngineException e) {
+                        versionConflict = true;
+                    }
+                } else {
+                    // index a document
+                    id = randomFrom(ids);
+                    ParsedDocument doc = testParsedDocument("test#" + id, id, "test", null, -1, -1, testDocumentWithTextField(), SOURCE, null);
+                    final Engine.Index index = new Engine.Index(newUid("test#" + id), doc,
+                        SequenceNumbersService.UNASSIGNED_SEQ_NO,
+                        rarely() ? 100 : Versions.MATCH_ANY, VersionType.INTERNAL,
+                        PRIMARY, 0, -1, false);
+                    try {
+                        initialEngine.index(index);
+                        indexedIds.add(id);
+                    } catch (VersionConflictEngineException e) {
+                        versionConflict = true;
+                    }
+                }
+                if (versionConflict == false) {
                     primarySeqNo++;
-                } catch (VersionConflictEngineException e) {
-
                 }
 
                 replicaLocalCheckpoint =
                     rarely() ? replicaLocalCheckpoint : randomIntBetween(Math.toIntExact(replicaLocalCheckpoint), Math.toIntExact(primarySeqNo));
                 initialEngine.seqNoService().updateLocalCheckpointForShard("primary", initialEngine.seqNoService().getLocalCheckpoint());
                 initialEngine.seqNoService().updateLocalCheckpointForShard("replica", replicaLocalCheckpoint);
 
+                // make sure the max seq no in the latest commit hasn't advanced due to more documents having been added;
+                // the first time the commit data iterable gets an iterator, the max seq no from that point in time should
+                // remain from any subsequent call to IndexWriter#getLiveCommitData unless the commit data is overwritten by a
+                // subsequent call to IndexWriter#setLiveCommitData.
+                assertThat(
+                    initialEngine.seqNoService().getMaxSeqNo(),
+                    // its possible we haven't indexed any documents yet, or its possible that right after a commit, a version conflict
+                    // exception happened so the max seq no was not updated, so here we check greater than or equal to
+                    initialEngine.seqNoService().getMaxSeqNo() != SequenceNumbersService.NO_OPS_PERFORMED || versionConflict ?
```

Comment:
I'm not sure we need this extra check? what were you aiming at testing ? feels like it's for the time we had caching?